eventhandlers.go

```diff
@@ -126,27 +110,34 @@ func HandleDeploymentFinishedEvent(myKeptn *keptn.Keptn, incomingEvent cloudeven
 	}
 	verdict = strings.Trim(verdict, `'"`)
 	log.Println("ChaosExperiment Verdict: " + verdict)
-	if verdict == "Pass" {
-		verdict = "pass"
-	} else {
-		verdict = "fail"
-	}
+
+	testResult := keptnv2.ResultFailed
+
+	// check verdict
 	log.Println("Final Result: " + verdict)
 
-	// Send Test Finished Event
-	if SendTestsFinishedEvent == "true" {
-		return myKeptn.SendTestsFinishedEvent(&incomingEvent, "", "", startTime, verdict, nil, "litmus-service")
+	if verdict == "Pass" {
+		testResult = keptnv2.ResultPass
 	}
-	return nil
+
+	// send tests.finished
+	_, err = myKeptn.SendTaskFinishedEvent(&keptnv2.EventData{
+		Status:  keptnv2.StatusSucceeded,
+		Result:  testResult,
+		Message: "Chaos tests finished",
+	}, ServiceName)
+
+	return err
 }
 
-//
-// Handles TestsFinishedEventType = "sh.keptn.events.tests-finished"
-//
-func HandleTestsFinishedEvent(myKeptn *keptn.Keptn, incomingEvent cloudevents.Event, data *keptn.TestsFinishedEventData) error {
+// HandleTestFinished handles test.finished event
+func HandleTestFinished(myKeptn *keptnv2.Keptn, incomingEvent cloudevents.Event, data *keptnv2.TestFinishedEventData) error {
 	log.Printf("Handling Tests Finished Event: %s", incomingEvent.Context.GetID())
 
-	// potential improvement: check if the Test-finished event is coming from Litmus service and ignore if so
+	if incomingEvent.Source() == ServiceName {
+		// skip test.finished, it has been send out by litmus service
+		return nil
+	}
 
```

Comment:
@ksatchit @jetzlstorfer To stay compatible with the Keptn 0.7.x implementation I am reacting on a test.finished event here. However, I'm not sure if we actually should stop chaos engine when a test.finished from another service comes in.

I'll leave that up to you if that makes sense or not.

Because of us canceling the chaos engine here, final result/verdict is an empty string (instead of success/pass).

Reply:
I double checked. I need to react on test.finished of jmeter, and stop chaos engine, else it would run indefinately.

However, if I stop it using kubectl delete, I end up getting an error, see screenshot below:
![image](https://user-images.githubusercontent.com/56065213/109010842-93beaa80-76b0-11eb-8319-d8a08ce2eaae.png)

Reply:
**Just to restate current behavior:** When an in-flight chaosengine is deleted upon receiving test.finished event from jmeter service (chaos still in progress as CHAOS_DURATION was set to a value higher than the jmeter run), we don't have a valid verdict in the chaosresult, leading to a parsing failure and hence test result is fail. If the jmeter runs for longer we delete a "completed chaosengine" by which time the verdict in the chaosresult might be pass/fail depending upon exit criteria for the experiment. 

In upcoming versions of litmus (this should already be available since 1.13 I suppose), deletion of an in-flight experiment will be treated as an "Abort" operation, with the verdict on the chaosresult updated to "Stopped". This is implemented so - as the experiment hypothesis will not be fully validated by this point. I can change/update the version of litmus dependencies from the current 1.9.1. 

I see the below set of possible values on Results: https://github.com/keptn/go-utils/blob/master/pkg/lib/v0_2_0/result.go

An initial thought at this point is to equate Abort status to Warning with a corresponding Message as it is technically not a confirmed failure. SLO checks at the quality gate will anyway determine that.

Reply:
Thank you so much @christian-kreuzberger-dtx for taking this up. Much appreciated.

Reply:
> chaos still in progress as CHAOS_DURATION was set to a value higher than the jmeter run

Would it make sense to introduce a loop and wait a couple of seconds, just in case Chaos finishes a couple of seconds after JMeter?

> I can change/update the version of litmus dependencies from the current 1.9.1.
I'll take a look at that

> An initial thought at this point is to equate Abort status to Warning with a corresponding Message as it is technically not a confirmed failure. SLO checks at the quality gate will anyway determine that.

Warning seems fair.

Reply:
I've left it at 1.9.1, but I've updated the workflow to send a warning.